aiida_lammps/workflows/base.py

```diff
@@ -0,0 +1,351 @@
+import re
+
+from aiida import orm
+from aiida.common import AttributeDict, InputValidationError
+from aiida.engine import (
+    BaseRestartWorkChain,
+    ProcessHandlerReport,
+    process_handler,
+    while_,
+)
+import numpy as np
+
+from aiida_lammps.calculations.base import LammpsBaseCalculation
+from aiida_lammps.utils import RestartTypes
+
+
+class LammpsBaseWorkChain(BaseRestartWorkChain):
+    """Base workchain for calculations using LAMMPS"""
+
+    _process_class = LammpsBaseCalculation
+
+    @classmethod
+    def define(cls, spec):
+        """Define the process specification."""
+        # yapf: disable
+        super().define(spec)
+        spec.expose_inputs(LammpsBaseCalculation, namespace='lammps')
+        spec.input(
+            "store_restart",
+            valid_type=orm.Bool,
+            default=lambda: orm.Bool(False),
+            required=False,
+            help="""
+            Whether to store the restartfile in the repository.
+            """,
+        )
+        spec.expose_outputs(LammpsBaseCalculation)
+        spec.outline(
+            cls.setup,
+            cls.validate_parameters,
+            while_(cls.should_run_process)(
+                cls.run_process,
+                cls.inspect_process,
+            ),
+            cls.results,
+        )
+
+        spec.exit_code(
+            300,
+            "ERROR_UNRECOVERABLE_FAILURE",
+            message="""
+            The calculation failed with an unidentified unrecoverable error.
+            """,
+        )
+        # yapf: enable
+
+    def setup(self):
+        """Call the ``setup`` of the ``BaseRestartWorkChain`` and create the inputs dictionary in ``self.ctx.inputs``.
+
+        This ``self.ctx.inputs`` dictionary will be used by the ``BaseRestartWorkChain`` to submit the calculations
+        in the internal loop.
+
+        The ``parameters`` and ``settings`` input ``Dict`` nodes are converted into a regular dictionary and the
+        default namelists for the ``parameters`` are set to empty dictionaries if not specified.
+        """
+        super().setup()
+        self.ctx.inputs = AttributeDict(
+            self.exposed_inputs(LammpsBaseCalculation, "lammps")
+        )
+        self.ctx.inputs.settings = (
+            self.ctx.inputs.settings.get_dict() if "settings" in self.ctx.inputs else {}
+        )
+
+    def validate_parameters(self):
+        """Validate the parameters for the workchain"""
+
+        if "store_restart" in self.ctx.inputs:
+            self.ctx.inputs.settings.store_restart = self.ctx.inputs.store_restart.value
+
+        if "script" not in self.ctx.inputs and any(
+            key not in self.ctx.inputs
+            for key in ["parameters", "potential", "structure"]
+        ):
+            raise InputValidationError(
+                "Unless `script` is specified the inputs `structure`, "
+                "`potential` and  `parameters` have to be specified."
+            )
```

Comment:
This validation is already done by `LammpsBaseCalculation` is it not? In that case you don't have to do it here. If you expose `LammpsBaseCalculation` its validators are called when you launch `LammpsBaseWorkChain`

Reply:
Aha, okay so when exposing the inputs the validators are called also? Okay that means that this is unnecessary indeed.

Reply:
Exactly. This was necessary, because otherwise the exposing functionality as a way to compose workchains together would be pretty much useless, since having to copy all validation each time would be untenable.